Onboarding: Services step; Newsletters wizard

adekbadek opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

Adds the Services step of the onboarding process, as well as a Newsletters wizard. 

### How to test the changes in this Pull Request:

1. Switch `newspack-newsletters` to [`add/update-settings` branch](https://github.com/Automattic/newspack-newsletters/pull/407)
2. Visit the Setup wizard, Services step (`/wp-admin/admin.php?page=newspack-setup-wizard#/services`)
3. Observe four sections, which can be toggled
4. Edit the first section ("Reader Revenue"), observe the settings are reflected in the Reader Revenue Wizard. Edit it there, too, and observe the settings updated when visiting the Services step again
5. Same with the next section ("Newsletters"), which maps to the "Authoring" part of the Engagement wizard
6. Observe that the last two sections (ad related) are mutually exclusive
7. Edit the ad related sections, observe changes reflected in the Advertising wizard
8. Visit the Engagement wizard, change the settings of the Newsletters and observe them being reflected in the setting of the Newspack Newsletters plugin 
9. Deactivate the Newspack Newsletters plugin. Observe the Engagement wizard displays a plugin card instead of the settings interface

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

thomasguillot reviewed (approved):
I made a bunch of visual changes

dkoo commented on assets/wizards/engagement/views/newsletters/index.js:
```diff
@@ -1,91 +1,225 @@
+			path: '/newspack/v1/wizard/newspack-engagement-wizard/newsletters',
+		} ).then( res => {
+			setAllProviders(
+				uniq( res.settings.reduce( ( acc, setting ) => [ ...acc, setting.provider ], [] ) ).filter(
```
When I visit the Engagement > Newsletters wizard, I only see Mailchimp as an option despite the `newsletters` endpoint returning the correct options.

<img width="1843" alt="Screen Shot 2021-02-23 at 1 23 54 PM" src="https://user-images.githubusercontent.com/2230142/108903541-a49cec80-75da-11eb-91e6-848825bd6c26.png">

dkoo reviewed (changes requested):
A few things I saw (running without the Gutenberg plugin, in case that makes a difference):

1. On the Services onboarding step, Reader Revenue section: If going through the steps on a site that had previously been set up with NRH set as the Reader Revenue platform, the Reader Revenue settings for Donation and Payment Gateway won't be displayed here despite existing in the database. Only after switching back to the "Newspack" platform in the Reader Revenue wizard did these setup fields pre-populate.
2. Newsletters section: the **Mailchimp API Key** field loses focus after each keystroke.
3. Google Ad Manager section: If going through setup steps on a new site in Debug mode, the **Network Code** field displays `[object Object]` at first:

<img width="1047" alt="Screen Shot 2021-02-23 at 11 14 22 AM" src="https://user-images.githubusercontent.com/2230142/108888363-9a71f280-75c8-11eb-9608-ae50537c4fa1.png">

4. If you enable several sections and then click the "Connect Google AdSense" link to configure settings and come back via the "Back to Newspack" button, all sections will appear disabled/collapsed on page load, and any values you had previously entered in other sections will be lost. Can we use some `localStorage` or URL params to remember which sections had been previously expanded and the values entered? I feel like because part of the guided flow to set up AdSense forces you to leave the page, this could cause frustration if the user had already filled out several other sections.
5. Clicking "Continue" after filling out all sections resulted in a 500 error for me on both a new JN instance and my local environment, despite having https://github.com/Automattic/newspack-newsletters/pull/404 checked out and active. Also seeing the same error when I try to click "save" from the Engagement > Newsletters wizard page. Error from my logs:

```
PHP Fatal error:  Uncaught Error: Call to undefined method Newspack_Newsletters_Settings::update_settings() in /Users/dkoo/Local Sites/newspack/app/public/wp-content/plugins/newspack-plugin/includes/configuration_managers/class-newspack-newsletters-configuration-manager.php:64
Stack trace:
#0 /Users/dkoo/Local Sites/newspack/app/public/wp-content/plugins/newspack-plugin/includes/wizards/class-setup-wizard.php(286): Newspack\Newspack_Newsletters_Configuration_Manager->update_settings(Array)
#1 /Users/dkoo/Local Sites/newspack/app/public/wp-includes/rest-api/class-wp-rest-server.php(1160): Newspack\Setup_Wizard->api_update_services(Object(WP_REST_Request))
#2 /Users/dkoo/Local Sites/newspack/app/public/wp-includes/rest-api/class-wp-rest-server.php(1007): WP_REST_Server->respond_to_request(Object(WP_REST_Request), '/newspack/v1/wi...', Array, NULL)
#3 /Users/dkoo/Local Sites/newspack/app/public/wp-includes/rest-api/class-wp-rest-server.php(440): WP_REST_Server->dispatch(Object(WP_REST_Request))
#4 /Users/dkoo/Local Sites/newspack in /Users/dkoo/Local Sites/newspack/app/public/wp-content/plugins/newspack-plugin/includes/configuration_managers/class-newspack-newsletters-configuration-manager.php on line 64
```

6. When I visit the Engagement > Newsletters wizard after deactivating the Newsletters plugin, I see the action card to activate the plugin. But after clicking "activate" and waiting for the plugin to become active, the action card just remains in a success state like the below. You have to either refresh the page or presumably hit "save" to see the settings after this (although clicking save for me resulted in the 500 error as described above). Would it be possible to automatically refresh the page or reset the view state on success?

<img width="1047" alt="Screen Shot 2021-02-23 at 1 13 30 PM" src="https://user-images.githubusercontent.com/2230142/108902198-0a887480-75d9-11eb-90dd-5cac04904518.png">

adekbadek commented:
> On the Services onboarding step, Reader Revenue section: If going through the steps on a site that had previously been set up with NRH set as the Reader Revenue platform, the Reader Revenue settings for Donation and Payment Gateway won't be displayed here despite existing in the database. 

Added some explanation text if the site is set up with NRH in 4eca7af171adb4d183994ab65febd323b264518d. Note that the Setup Wizard will not handle NRH setup - its aim is to just to provide the defaults.

> Newsletters section: the Mailchimp API Key field loses focus after each keystroke.

Fixed in 4308871ddf22757e15eebf6f254c96f431a93163

> Google Ad Manager section: If going through setup steps on a new site in Debug mode, the Network Code field displays [object Object] at first:

Added missing/misconfigured Ads plugin handling in d90364447f5cc8a6010bd55625eaf6ebc5f25dbd

> When I visit the Engagement > Newsletters wizard after deactivating the Newsletters plugin, I see the action card to activate the plugin. But after clicking "activate" and waiting for the plugin to become active, the action card just remains in a success state like the below.

I don't think we need to handle that scenario. Part of the onboarding flow consists of activating the Newsletters plugin - deactivating it mid-flow is an extreme edge case.

> If you enable several sections and then click the "Connect Google AdSense" link to configure settings and come back via the "Back to Newspack" button, all sections will appear disabled/collapsed on page load, and any values you had previously entered in other sections will be lost. 

Good catch. Changed to handle AdSense configuration in a new tab in 5997174571f643841697c81942f662c1b11a36b2. I know it's not the best UX, but we'll probably be dropping AdSense in favor of Ad Manager in some time anyway.

> When I visit the Engagement > Newsletters wizard, I only see Mailchimp as an option despite the newsletters endpoint returning the correct options.

> Clicking "Continue" after filling out all sections resulted in a 500 error 

Sorry for that, I forgot to add a commit to https://github.com/Automattic/newspack-newsletters/pull/404 - added as https://github.com/Automattic/newspack-newsletters/pull/407

dkoo reviewed (changes requested):
> I don't think we need to handle that scenario. Part of the onboarding flow consists of activating the Newsletters plugin - deactivating it mid-flow is an extreme edge case.

Oh, I didn't mean this happens mid-flow—I meant after completing onboarding, if you visit Engagement > Newsletters while the Newsletters plugin is deactivated. It's not a huge deal, but if it's an easy change to update the view state after activating, I think it'd be a nice improvement. Not a blocker since it's probably still somewhat of an edge case.

All other changes look good, just needs a conflict resolution and possibly a tweak for the above, and this is 👍

adekbadek commented:
> I meant after completing onboarding, if you visit Engagement > Newsletters while the Newsletters plugin is deactivated.

Ah, got you. Added in e79156cfae7638443e34c8c5b1719f74e4243d13

dkoo reviewed (approved):
Ah, looks good now!
